qualtran/bloqs/mod_arithmetic/mod_addition.py

```diff
@@ -247,5 +131,201 @@ def build_composite_bloq(self, bb: 'BloqBuilder', x: Soquet, y: Soquet) -> Dict[
+            return
+        if not 1 <= value <= 2**self.bitsize:
+            raise ValueError(f"mod: {value} must be between [1, {2 ** self.bitsize}].")
+
+    @cached_property
+    def signature(self) -> Signature:
+        if self.cvs:
+            return Signature(
+                [
+                    Register('ctrl', QBit(), shape=(len(self.cvs),)),
+                    Register('x', QUInt(self.bitsize)),
+                ]
+            )
+        return Signature([Register('x', QUInt(self.bitsize))])
+
+    def _classical_unctrled(self, target_val: int):
+        if target_val < self.mod:
+            return (target_val + self.add_val) % self.mod
+        return target_val
+
+    def on_classical_vals(
+        self, *, x: int, ctrl: Optional[int] = None
+    ) -> Dict[str, 'ClassicalValT']:
+        out = self._classical_unctrled(x)
+        if self.cvs:
+            assert ctrl is not None
+            if ctrl == int(''.join(str(x) for x in self.cvs), 2):
+                return {'ctrl': ctrl, 'x': out}
+            else:
+                return {'ctrl': ctrl, 'x': x}
+
+        assert ctrl is None
+        return {'x': out}
+
+    def __pow__(self, power: int) -> 'AddK':
+        return AddK(self.bitsize, self.mod, k=self.add_val * power, cvs=self.cvs)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Add(QUInt(self.bitsize), QUInt(self.bitsize)), 5)}
+
+
+@bloq_example
+def _mod_add_k() -> ModAddK:
+    n, m, k = sympy.symbols('n m k')
+    mod_add_k = ModAddK(bitsize=n, mod=m, add_val=k)
+    return mod_add_k
+
+
+@bloq_example
+def _mod_add_k_small() -> ModAddK:
+    mod_add_k_small = ModAddK(bitsize=4, mod=7, add_val=1)
+    return mod_add_k_small
+
+
+@bloq_example
+def _mod_add_k_large() -> ModAddK:
+    mod_add_k_large = ModAddK(bitsize=64, mod=500, add_val=23)
+    return mod_add_k_large
+
+
+_MOD_ADD_K_DOC = BloqDocSpec(
+    bloq_cls=ModAddK, examples=[_mod_add_k, _mod_add_k_small, _mod_add_k_large]
+)
+
+
+@frozen
+class CModAddK(Bloq):
+    """Perform x += k mod m for constant k, m and quantum x.
+
+    Args:
+        k: The integer to add to `x`.
+        mod: The modulus for the addition.
+        bitsize: The bitsize of the `x` register.
+
+    Registers:
+        ctrl: The control bit
+        x: The register to perform the in-place modular addition.
+    """
+
+    k: Union[int, sympy.Expr]
+    mod: Union[int, sympy.Expr]
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('ctrl', QBit()), Register('x', QUInt(self.bitsize))])
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        k = ssa.new_symbol('k')
+        return {(AddK(k=k, bitsize=self.bitsize).controlled(), 5)}
+
+    def short_name(self) -> str:
+        return f'x += {self.k} % {self.mod}'
+
+
+@frozen
+class CtrlScaleModAdd(Bloq):
```

Comment:
There's a naming convention mismatch here right? You're using CModAddK above and Ctrl... here.

Reply:
This one is just a straight move with no changes. It's on the list for a re-name but lower priority since it's a bespoke primitive used by the modular exponentiation factoring